components/hab/src/command/plan/render.rs

```diff
@@ -0,0 +1,176 @@
+use crate::error::Result;
+
+pub fn start(
+    ui: &mut UI,
+    template_path: String,
+    default_toml_path: String,
+    user_toml_path: Option<String>,
+    mock_data_path: Option<String>,
+    print: bool,
+    no_render_dir: bool,
+    render_dir: String,
+    quiet: bool,
+) -> Result<()> {
+    // Strip the file name out of our passed template
+    let file_name = match Path::new(&template_path).file_name() {
+        Some(name) => name.to_str().clone().unwrap(),
+        None => panic!(format!("Something went wrong getting filename of {}", &template_path)),
+    }; 
+
+    if !(quiet) {
+      ui.begin(format!("Rendering: {} into: {}/ as: {}", template_path, render_dir, file_name))?;
+      ui.br()?;
+    }
+
+    // read our template from file
+    let template = read_to_string(&template_path)
+        .expect(&format!("something went wrong reading: {}", template_path)); 
+
+    // create a "data" json struct
+    let mut data: Json = serde_json::from_str("{}").unwrap();
+
+    if !(quiet) {
+        // import default.toml values, convert to JSON
+        ui.begin(format!("Importing default.toml: {}", &default_toml_path))?;
+    }
+
+    // we should always have a default.toml, would be nice to "autodiscover" based on package name,
+    // for now assume we're working in the plan dir if --default-toml not passed
+    let default_toml = read_to_string(&default_toml_path)
+        .expect(&format!("Something went wrong reading: {}", &default_toml_path));
+    // merge default into data struct
+    merge(&mut data, toml_to_json(&default_toml));
+
+    // import default.toml values, convert to JSON
+    // ui.begin(format!("Importing user.toml: {}", &user_toml_path));
+    let user_toml = match user_toml_path {
+        Some(path) => {
+            if !(quiet) {
+              // print helper message, maybe only print if '--verbose'? how?
+              ui.begin(format!("Importing user.toml: {}", path.to_string()))?;
+            }
+            read_to_string(path.to_string())
+                .expect(&format!("Something went wrong reading: {}", path.to_string()))
+        },
+        None => "".to_string(),
+    };
+    // merge default into data struct
+    merge(&mut data, toml_to_json(&user_toml));
+
+    // read mock data if provided
+    let mock_data = match mock_data_path {
+        Some(path) => {
+            if !(quiet) {
+                // print helper message, maybe only print if '--verbose'? how?
+                ui.begin(format!("Importing override file: {}", path.to_string()))?;
+            }
+            read_to_string(path.to_string())
+              .expect(&format!("Something went wrong reading: {}", path.to_string()))
+        },
+        // return an empty json block if '--mock-data' isn't defined.
+        // this allows us to merge an empty JSON block
+        None => "{}".to_string(),
+    };
+    // merge mock data into data
+    merge(&mut data, serde_json::from_str(&mock_data).unwrap());
```

Comment:
Also, the use of `unwrap()` anywhere but example code discouraged because it causes a panic. There are some cases where it's OK to use, but those would be places where you've already determined by other means that it won't panic. A prime example of this would be [here](https://github.com/habitat-sh/habitat/pull/6114/files#r256617348)

Since we're parsing user-supplied data here, this operation could definitely fail (what if the user supplied a TOML file by mistake?).

Since you're in a function that already returns a `Result` type, you are free to use the [? operator](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator) instead of `unwrap()`.

(For completeness, you can also explicitly handle the error case if that makes sense, but I think `?` is fine here.)

Reply:
Hmm... when I use `?` here I get:

```
error[E0277]: the trait bound `error::Error: std::convert::From<serde_json::Error>` is not satisfied
   --> components/hab/src/command/plan/render.rs:100:22
    |
100 |     merge(&mut data, serde_json::from_str(&mock_data)?);
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::convert::From<serde_json::Error>` is not implemented for `error::Error`
```

I think his might be related to the other error thread?

Reply:
Yeah, see 
https://github.com/habitat-sh/habitat/pull/6114#discussion_r260959820

Reply:
:+1: